lib/vdc/BOVCollection.cpp

```diff
@@ -0,0 +1,368 @@
+    if (pos != std::string::npos) {    // We found the token
+        T lineValue;
+        _findTokenValue(line);
+        std::stringstream lineStream(line);
+
+        value.clear();
+        while (lineStream >> lineValue) { value.push_back(lineValue); }
+
+        if (value.size() != 3) {
+            value.clear();
+            std::string message = token + " must be a set of three values.";
+            SetErrMsg(message.c_str());
+            return -1;    // Bad token
+        }
+
+        if (lineStream.bad()) {
+            value.clear();
+            std::cout << "FAIL v " << token << std::endl;
+            std::string message = "Invalid value for " + token + " in BOV header file.";
+            SetErrMsg(message.c_str());
+            return -1;    // Bad token
+        }
+
+        if (verbose) {
+            std::cout << std::setw(20) << token << " ";
+            for (int i = 0; i < value.size(); i++) std::cout << value[i] << " ";
+            std::cout << std::endl;
+        }
+        return 1;    // we now have a token
+    }
+    return 0;    // we didn't find the token
+}
+
+void BOVCollection::_findTokenValue(std::string &line) const
+{
+    std::string delimiter = ": ";
+
+    size_t      pos = 0;
+    std::string token;
+    while ((pos = line.find(delimiter)) != std::string::npos) {
+        token = line.substr(0, pos);
+        line.erase(0, pos + delimiter.length());
+    }
+}
+
+size_t BOVCollection::_sizeOfFormat(DC::XType type) const
+{
+    switch (type) {
+    case DC::XType::INT32: return 4;
+    case DC::XType::FLOAT: return 4;
+    case DC::XType::DOUBLE: return 8;
+    default: return -1;
+    }
+}
+
+void BOVCollection::_swapBytes(void *vptr, size_t size, size_t n) const
+{
+    unsigned char *ucptr = (unsigned char *)vptr;
+    unsigned char  uc;
+    size_t         i, j;
+
+    for (j = 0; j < n; j++) {
+        for (i = 0; i < size / 2; i++) {
+            uc = ucptr[i];
+            ucptr[i] = ucptr[size - i - 1];
+            ucptr[size - i - 1] = uc;
+        }
+        ucptr += size;
+    }
+}
+
+template<class T> int BOVCollection::ReadRegion(const std::vector<size_t> &min, const std::vector<size_t> &max, T region)
+{
+    FILE *fp = fopen(_dataFile.c_str(), "rb");
+    if (!fp) {
+        SetErrMsg("Invalid file: %d", fp);
+        return -1;
+    }
+
+    size_t formatSize = _sizeOfFormat(_dataFormat);
+    if (formatSize < 0) {
+        SetErrMsg("Unspecified data format");
+        return (-1);
+    }
+
+    if (_dataSize.size() != 3) {
+        SetErrMsg("Invalid grid size (must be 3D)");
+        return (-1);
+    }
+    size_t numValues = _dataSize[0] * _dataSize[1] * _dataSize[2];
+
+    int  n = 1;
+    bool systemLittleEndian = *(char *)&n == 1 ? true : false;
+    bool dataLittleEndian = _dataEndian == "LITTLE" ? true : false;
+    bool needSwap = systemLittleEndian != dataLittleEndian ? true : false;
+
+    // Read a "pencil" of data along the X axis, one row at a time
```

Comment:
I believe this way of reading a file will result in an unnecessarily big number of file operation requests. It might be OK on a personal computer, but on a HPC with parallel filesystems, this practice will result in poor performance.

I suggest you read in the whole file using one read request.

Reply:
@shaomeng raises an interesting point, but the API needs to support the ability to read a **subset** of the data. Doing as Sam suggests would entail reading more data than necessary (in some cases a lot more data), and then copying the requested subset to the destination buffer. This might result in better performance, but modern operating systems are pretty smart about handling buffered I/O (and setvbuf() can be used for fine tuning if needed). I'd suggest convincing oneself that there is a problem before introducing an optimization that might not provide any benefit, at the cost of more complicated code.

Reply:
@clyne you're asking scott to do something (reading a subset of a data file) that the VisIT BOV specification doesn't support. Scott will need to come up with a scheme to augment the BOV specification to support it. I don't think we should go down that route. (I think we should stick to what BOV already supports which we are supporting a only subset at this point.)

Reply:
@shaomeng I'm not sure I understand your comment. The BOV specification is for a file format, not an API. AFAIK it doesn't say anything about how the data should be accessed, only how the data are organized.  In any event, the VAPOR API requires the ability to read a subset of the data. Scott has already implemented this. However, you made the comment that his implementation was inefficient. I'm not sure that it is, and suggested that we not complicate the implementation prematurely for optimization purposes until we know there is a problem.

Reply:
I see what you mean. I'm OK to deal with it later if there's a performance issue.